frontend/src/reducers/user.js

```diff
@@ -0,0 +1,140 @@
+      headers: { 'Content-Type': 'application/json' },
+    })
+      .then((res) => {
+        //Failed
+        if (!res.ok) {
+          throw new Error('Could not create account')
+        }
+        //Success
+        return res.json()
+      })
+      .then((json) => {
+        // Save the login info
+        dispatch(
+          user.actions.setAccessToken({
+            accessToken: json.accessToken,
+          })
+        )
+        dispatch(user.actions.setUserName({ userName: json.userName }))
+      })
+      .catch((err) => {
+        dispatch(user.actions.setErrorMessage({ errorMessage: err.toString() }))
+      })
+  }
+}
+
+//_______Login
+export const login = (name, email, password) => {
+  const LOGIN_URL = 'https://the-secret.herokuapp.com/login'
+  return (dispatch) => {
+    fetch(LOGIN_URL, {
+      method: 'POST',
+      body: JSON.stringify({ name, email, password }),
+      headers: { 'Content-Type': 'application/json' },
+    })
+      .then((res) => {
+        //Success
+        if (res.ok) {
+          return res.json()
+        }
+        // Not OK - failed
+        throw new Error('Unable to sign in. Please check your username and password are correct')
+      })
+      .then((json) => {
+        // Save the login info
+        dispatch(
+          user.actions.setAccessToken({
+            accessToken: json.accessToken,
+          })
+        )
+        dispatch(user.actions.setUserName({ userName: json.userName }))
+      })
+      .catch((err) => {
+        dispatch(logout())
+        dispatch(user.actions.setErrorMessage({ errorMessage: err.toString() }))
+      })
+  }
+}
+
+//_______Secret message
+export const getSecretMessage = () => {
+  const USERS_URL = 'https://the-secret.herokuapp.com/users'
+  return (dispatch, getState) => {
+    const accessToken = getState().user.login.accessToken
+    const userId = getState().user.login.userId
+    fetch(`${USERS_URL}/${userId}/secret`, {
+      method: 'GET',
+      // Include the accessToken to get the protected endpoint
+      headers: { Authorization: accessToken },
+    })
+      .then((res) => {
+        if (res.ok) {
+          return res.json()
+        }
+        throw new Error('Could not get information. Make sure you are logged in and try again.')
+      })
+      .then((json) => {
+        dispatch(
+          user.actions.setSecretMessage({ secretMessage: JSON.stringify(json).slice(1, -1) })
+        )
+      })
+      .catch((err) => {
+        dispatch(user.actions.setErrorMessage({ errorMessage: err.toString() }))
+      })
+  }
+}
+
+export const logout = () => {
+  return (dispatch) => {
+    dispatch(user.actions.setSecretMessage({ secretMessage: null }))
+    dispatch(user.actions.setErrorMessage({ errorMessage: null }))
+    dispatch(user.actions.setAccessToken({ accessToken: null }))
+    dispatch(user.actions.setUserName({ userName: null }))
+    localStorage.removeItem('accessToken')
+    localStorage.removeItem('userName')
```

Comment:
Could you not just write 'localStorage.clear()' here?

Reply:
It is not a good idea to use `localStorage.clear()` because when you have a bigger app you can want to store more data in the local storage. Some of them, like shopping cards, should stay when you logging out 😸